stdlib/condition.mli

```diff
@@ -13,41 +13,112 @@
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
+
+   Assuming that [D] is a shared data structure
+   protected by the mutex [m],
+   and assuming that [c] is a condition variable
+   associated with the mutex [m]
+   and with the property [P],
+   a typical usage scenario is as follows:
    {[
      Mutex.lock m;
-     while (* some predicate P over D is not satisfied *) do
+     while (* the property P over D is not satisfied *) do
        Condition.wait c m
      done;
      (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     if (* the property P over D is now satisfied *) then Condition.signal c;
      Mutex.unlock m
    ]}
+
+   {!signal} and {!broadcast} should be called only when the mutex [m]
+   associated with the condition variable [c] is locked.
+   Otherwise, there would be a risk that some other thread is currently
+   inside a critical section, has determined that the property [P]
+   associated with [c] is false, but has not yet called [wait]. In such a
+   case, the signal would be lost, leading to a potential deadlock.
 *)
 
 type t
 (** The type of condition variables. *)
 
 val create : unit -> t
-(** Return a new condition variable. *)
+(**[create()] creates and returns a new condition variable.
+   This condition variable should be associated (in the programmer's mind)
+   with a certain mutex [m], and should be used only when the mutex [m] is
+   locked. *)
 
 val wait : t -> Mutex.t -> unit
-(** [wait c m] atomically unlocks the mutex [m] and suspends the
-   calling process on the condition variable [c]. The process will
-   restart after the condition variable [c] has been signalled.
-   The mutex [m] is locked again before [wait] returns. *)
+(**The call [wait c m] is permitted only if [m] is the mutex associated
+   with the condition variable [c], and only if [m] is currently locked.
+   This call atomically unlocks the mutex [m] and suspends the
+   current thread on the condition variable [c]. This thread can
+   later be woken up after the condition variable [c] has been signaled
+   via {!signal} or {!broadcast}; however, it can also be woken up for
+   no reason. The mutex [m] is locked again before [wait] returns. One
+   cannot assume that the property [P] associated with the condition
+   variable [c] holds when [wait] returns; one must explicitly test
+   whether [P] holds after calling [wait]. *)
 
 val signal : t -> unit
-(** [signal c] restarts one of the processes waiting on the
-   condition variable [c]. *)
+(**[signal c] wakes up one of the threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
+   associated with [c] is currently locked. It should be the case that
+   the property [P] associated with the condition variable [c] is true
+   when [signal] is called. *)
```

Comment:
Sentence "It should be ... is called" is not clear. I fact as the above comment mandates re-checking [P] after having returned from suspension, there is nothing dangerous in signaling witthout checking [P]. Moreover when implementing a concurrent FIFO, tere are two correct alternatives for signaling the threads thar wait for the queue to become non-empty:
1. 'put' calls {! broadcast} when it puts some element in an empty queue.
2. 'put' calls {! signal} in all situations.

Herlihy and Shavit' "The art of multiprocessor programming" uses technique 2 in its FIFO implementation (Sec. 8.2.2)

Reply:
I have removed the indication that `P` should be true when `signal` or `broadcast` is called.